test/test_cuda.py

```diff
@@ -3012,6 +3012,35 @@ def test_graph_record_stream(self):
         # dummy allocation triggers process_events, Hopefully successfully processes b's end-of-life event.
         torch.zeros((3,), device="cuda")
 
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs"
+    )
+    def test_graph_record_stream_no_leak(self):
+        torch.cuda.empty_cache()
+        torch.cuda.synchronize()
+        base_mem = torch.cuda.memory_allocated()
+
+        g = torch.cuda.CUDAGraph()
+        s_main = torch.cuda.Stream()
+        s_other = torch.cuda.Stream()
+
+        with torch.cuda.stream(s_main):
+            g.capture_begin()
+            x = torch.empty(64 * 1024 * 1024 // 4, device="cuda")
+            s_other.wait_stream(s_main)
+            with torch.cuda.stream(s_other):
+                y = x + 1
+                y.record_stream(s_other)
+                del y
+            s_main.wait_stream(s_other)
+            del x
```

Comment:
The problem is that we OOM before the graph capture is done. Probably put these lines in a for loop with 100 iterations or so and allocate like ~10GB each iteration and I think it would OOM despite deleting x. Also call `record_stream` on `x` in the other stream.

Reply:
All of the changes  to the allocator in this PR are in `endAllocateToPool` so by definition that won't affect anything in the middle of a capture ;)

Reply:
I think if done correctly there shouldn't be much perf overhead. We need to keep track of producer and consumer streams in the autograd engine per tensor. Once the tensor is deleted, we record an event on the consumer stream and make the producer stream wait for that event

Reply:
@soulitzer The problem is that in a graph capture record_stream semantics change into "this tensor cannot be free'd until the capture is complete" which is effectively a memory leak. This issue has no easy workaround due to `record_stream` relying on a host-synchronization (illegal during graph capture) to track when the tensor is safe to free. The above PR attempts to trade between potential performance (forcing the producer stream to wait for the consumer stream to be done) and memory overhead.